### PR TITLE
Add the alpha dropout method

### PR DESCRIPTION
**Pull request by BoltzmannZhaung**
Dear Post Dr.Bourgin,
I wanna give a perfect method code, however , unfortunately may be it can not work well (I have not test the method).So.....uh~~  may I ask your help to finish it, then get the honor that be a contributor of this great project (5k+ stars)....
THANK YOU!

### CODE REVIEW

**Comment by WuZhuoran** [comment]
Maybe you can refer keras's `Alpha Dropout` [here](https://github.com/keras-team/keras/blob/master/tests/keras/layers/noise_test.py) to create unit test.

**Comment by BoltzmannZhaung** [comment]
How fantastic!!! Thank you~

**Comment by ddbourgin** [comment]
Hi @BoltzmannZhaung - thanks for doing this, on first glance it looks great :) Will take a closer look soon.

**Comment by BoltzmannZhaung** [comment]
thank u~!

File: neural_nets/wrappers/wrappers.py [hunk 1]
```diff
@@ -142,11 +142,59 @@ def backward(self, dLdy):
...
+    def forward(self, X):
+        alpha = 1.6732632423543772848170429916717
+        scale = 1.0507009873554804934193349852946
+        alpha_p = -alpha * scale
+        if self.trainable:
+            scaler = 1.0 - self.p
+            a = ((1 - self.p) * (1 + self.p * alpha_p ** 2)) ** -0.5
+            b = -a * alpha_p * self.p
+
+            keep = self._greaterOrEqual(np.random.uniform(0,1,X.shape))
```
Comment (ddbourgin) [inline neural_nets/wrappers/wrappers.py]: Does this run? It looks like you don't pass a `y` value to your `_greaterOrEqual` method.

In general, I don't think you need to create a separate method. A one-liner is still pretty clean (IMO): `np.random.rand(*X.shape) >= self.p`.

File: neural_nets/wrappers/wrappers.py [hunk 2]
```diff
@@ -142,11 +142,59 @@ def backward(self, dLdy):
         dLdy *= 1.0 / (1.0 - self.p)
         return self._base_layer.backward(dLdy)
 
+class AlphaDropout(WrapperBase):
+    def __init__(self, wrapped_layer, p):
+        """
+       Alpha Dropout is a `Dropout` that aim at keeping mean and variance to
```
Comment (ddbourgin) [inline neural_nets/wrappers/wrappers.py]: It looks like this description is copied from the Tensorflow documentation. I think it's better if we have documentation be more explicit + verbose for a project like this, since the expectation is that these will be "reference" implementations for people to use when trying to understand how the different model components work.

With that in mind, this should be rewritten. I'm happy to help proofread and revise grammar as you'd like :-)

File: neural_nets/wrappers/wrappers.py [hunk 3]
```diff
@@ -142,11 +142,59 @@ def backward(self, dLdy):
...
+        alpha_p = -alpha * scale
+        if self.trainable:
+            scaler = 1.0 - self.p
+            a = ((1 - self.p) * (1 + self.p * alpha_p ** 2)) ** -0.5
+            b = -a * alpha_p * self.p
+
+            keep = self._greaterOrEqual(np.random.uniform(0,1,X.shape))
+            x = X * keep + alpha_p * (1 - keep)
+            X = a * x + b
+        self._wrapper_derived_variables["AlphaDropout_mask"] = keep
```
Comment (ddbourgin) [inline neural_nets/wrappers/wrappers.py]: This is inconsistent with `dropout_mask` in line 165. To maintain consistency with the regular `Dropout` wrapper, I'd recommend changing this to `dropout_mask` as well.

File: neural_nets/wrappers/wrappers.py (hunk 1 above)
Comment (ddbourgin) [reply neural_nets/wrappers/wrappers.py]: More generally, this function is *very* similar to the Keras implementation [here](https://github.com/keras-team/keras/blob/ed07472bc5fc985982db355135d37059a1f887a9/keras/layers/noise.py#L144). 

I want to stress that all code in this repo should be your own work, or should properly cite the code it's based on.

File: neural_nets/wrappers/wrappers.py [hunk 4]
```diff
@@ -142,11 +142,59 @@ def backward(self, dLdy):
...
+        return scaler * self._base_layer.forward(X)
+
+    def backward(self, dLdy):
```
Comment (ddbourgin) [inline neural_nets/wrappers/wrappers.py]: This is not correct for `AlphaDropout`, as `X` is not scaled by `1 / 1-p` - it's a function of alpha, scale, and p now.

**Comment by ddbourgin** [comment]
Hi @BoltzmannZhaung - sorry for the delay. I just did a review pass and left some specific in-line comments. In addition to these, there are two general points I want to emphasize:

1. Please double check to make sure your code runs when you submit. I noticed at least one bug in your code that would easily have been caught if you tried running it on an example.

2. It's important that the code in this repo represent a sincere effort at original work. In particular, this means that:

    - It is not directly copied from other sources
    - If it is based off of an existing implementation, those implementations are cited in both the function  documentation and the PR itself.
 
I recognize that for simple functions it can be difficult to implement them cleanly in a way that doesn't  resemble existing implementations. This is okay, and I don't expect you to be able to write 100% unique code. What I want, however, is that your code represent *your* efforts, rather than someone else's :)

**Comment by BoltzmannZhaung** [comment]
@ddbourgin
Thanks for ur review and comments!  I will rewrite this method with the individual code immediately that represents my individual efforts. This repo which represents a sincere effort at the original work you have declared did inspire me with determination.
THANK YOU!
ps: uh... I give you a sincere apology for my previous bad work. for this time, I will try my best to finish it correctly and as original as possible.
oh~ also as soon as possible.
